src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionCommitManager.cs

```diff
@@ -0,0 +1,167 @@
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+using EditorCompletion = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.EditorImplementation
+{
+    internal sealed class CompletionCommitManager : IAsyncCompletionCommitManager
+    {
+        // We use an array instead of ImmutableArray here becuase we need to provide these characters to 
+        // IEnumerable<char> IAsyncCompletionCommitManager.PotentialCommitCharacters.
+        // Using ImmutableArray would lead to a regular boxing which is an unnecessary preformance issue.
+        private static readonly char[] s_commitChars = new char[] {
+            ' ', '{', '}', '[', ']', '(', ')', '.', ',', ':',
+            ';', '+', '-', '*', '/', '%', '&', '|', '^', '!',
+            '~', '=', '<', '>', '?', '@', '#', '\'', '\"', '\\'};
+
+        private static readonly EditorCompletion.CommitResult CommitResultUnhandled = new EditorCompletion.CommitResult(isHandled: false, EditorCompletion.CommitBehavior.None);
+
+        IEnumerable<char> IAsyncCompletionCommitManager.PotentialCommitCharacters => s_commitChars;
+
+        public bool ShouldCommitCompletion(IAsyncCompletionSession session, SnapshotPoint location, char typedChar, CancellationToken token)
+            => s_commitChars.Contains(typedChar);
+
+        public EditorCompletion.CommitResult TryCommit(IAsyncCompletionSession session, ITextBuffer buffer, EditorCompletion.CompletionItem item, char typeChar, CancellationToken token)
+        {
+            var document = buffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return CommitResultUnhandled;
+            }
+
+            var completionService = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(CompletionItemSource.RoslynItem, out var roslynItem))
+            {
+                // This isn't an item we provided (e.g. Razor). Let the editor handle it normally.
+                return CommitResultUnhandled;
+            }
+
+            if (!IsCommitCharacter(typeChar, roslynItem.Rules.CommitCharacterRules))
```

Comment:
i'm still a little confused.  is TryCommit called if ShouldCommitCompletion returns false?  

If not, then we have a problem.  Because someone may opt for a completion character that isn't in that big list.  So we'll return 'false', then not get called into here at all...

Reply:
We assume that ShouldCommitCompletion should return true for all possible characters that can lead to commit even ones could be overridden or added by overrides in Rules.

We could always return true in ShouldCommitCompletion but the current one seems to be some kind of optimization.

---
In reply to: [222074286](https://github.com/dotnet/roslyn/pull/29016#discussion_r222074286) [](ancestors = 222074286)